typed-protocols/src/Network/TypedProtocol/Driver/General.hs

```diff
@@ -0,0 +1,309 @@
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE TypeFamilies #-}
+{-# LANGUAGE RankNTypes #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE NamedFieldPuns #-}
+{-# LANGUAGE TypeInType #-}
+{-# LANGUAGE EmptyCase #-}
+
+-- | Actions for running 'Peer's with a 'Driver'
+--
+module Network.TypedProtocol.Driver.General (
+
+  -- * Introduction
+  -- $intro
+
+  -- * Driver interface
+  Driver(..),
+
+  -- * Normal peers
+  runPeerWithDriver,
+
+  -- * Pipelined peers
+  runPipelinedPeerWithDriver,
+  ) where
+
+import Data.Void (Void)
+
+import Network.TypedProtocol.Core
+import Network.TypedProtocol.Pipelined
+import Network.TypedProtocol.Codec (SomeMessage(..))
+
+import Control.Monad.Class.MonadSTM
+import Control.Monad.Class.MonadAsync
+
+
+-- $intro
+--
+-- A 'Peer' is a particular implementation of an agent that engages in a
+-- typed protocol. To actualy run one we need a source and sink for the typed
+-- protocol messages. These are provided by a 'Channel' and a 'Codec'. The
+-- 'Channel' represents one end of an untyped duplex message transport, and
+-- the 'Codec' handles conversion between the typed protocol messages and
+-- the untyped channel.
+--
+-- So given the 'Peer' and a compatible 'Codec' and 'Channel' we can run the
+-- peer in some appropriate monad. The peer and codec have to agree on
+-- the same protocol and role in that protocol. The codec and channel have to
+-- agree on the same untyped medium, e.g. text or bytes. All three have to
+-- agree on the same monad in which they will run.
+--
+-- This module provides drivers for normal and pipelined peers. There is
+-- very little policy involved here so typically it should be possible to
+-- use these drivers, and customise things by adjusting the peer, or codec
+-- or channel.
+--
+-- It is of course possible to write custom drivers and the code for these ones
+-- may provide a useful starting point. The 'runDecoder' function may be a
+-- helpful utility for use in custom drives.
+--
+
+
+--
+-- Driver interface
+--
+
+data Driver ps dstate m =
+        Driver {
+          sendMessage :: forall (pr :: PeerRole) (st :: ps) (st' :: ps).
+                         PeerHasAgency pr st
+                      -> Message ps st st'
+                      -> m ()
```

Comment:
It is specifically decoder state. It's just a generalisation of the existing trailing data. If we threaded state via the send side too then it would not work with the existing structure of the pipelined peer driver.